godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,480 @@
+        self.as_ref().to_string()
+    }
+
+    fn owner(&self) -> Gd<Object> {
+        self.as_ref().owner()
+    }
+
+    fn property_state(&self) -> Vec<(StringName, Variant)> {
+        self.as_ref().property_state()
+    }
+
+    fn language(&self) -> Gd<ScriptLanguage> {
+        self.as_ref().language()
+    }
+
+    fn refcount_incremented(&self) {
+        self.as_ref().refcount_incremented()
+    }
+
+    fn refcount_decremented(&self) -> bool {
+        self.as_ref().refcount_decremented()
+    }
+}
+
+struct ScriptInstanceData<T: ScriptInstance> {
+    inner: RefCell<T>,
+}
+
+pub fn create_script_instance<T: ScriptInstance>(rs_instance: T) -> GDExtensionScriptInstancePtr {
+    let gd_instance = crate::sys::GDExtensionScriptInstanceInfo {
+        set_func: Some(script_instance_info::set_func::<T>),
+        get_func: Some(script_instance_info::get_func::<T>),
+        get_property_list_func: Some(script_instance_info::get_property_list_func::<T>),
+        free_property_list_func: Some(script_instance_info::free_property_list_func::<T>),
+        // unimplemented until it's clear if it's needed
+        property_get_revert_func: None,
+        // unimplemented until it's clear if it's needed
+        property_can_revert_func: None,
+        get_owner_func: Some(script_instance_info::get_owner_func::<T>),
+        get_property_state_func: Some(script_instance_info::get_property_state_func::<T>),
+        get_script_func: Some(script_instance_info::get_script_func::<T>),
+        get_language_func: Some(script_instance_info::get_language_func::<T>),
+        // usage is unclear
+        get_fallback_func: None,
+        get_method_list_func: Some(script_instance_info::get_method_list_func::<T>),
+        get_property_type_func: Some(script_instance_info::get_property_type_func::<T>),
+        free_func: Some(script_instance_info::free_func::<T>),
+        free_method_list_func: Some(script_instance_info::free_method_list_func::<T>),
+        has_method_func: Some(script_instance_info::has_method_func::<T>),
+        call_func: Some(script_instance_info::call_func::<T>),
+        // deprecated by Godot
+        notification_func: None,
+        to_string_func: Some(script_instance_info::to_string_func::<T>),
+        refcount_decremented_func: Some(script_instance_info::refcount_decremented_func::<T>),
+        refcount_incremented_func: Some(script_instance_info::refcount_incremented_func::<T>),
+        is_placeholder_func: Some(script_instance_info::is_placeholder_func::<T>),
+        // usage is unclear
+        set_fallback_func: None,
+    };
+
+    let data = ScriptInstanceData {
+        inner: RefCell::new(rs_instance),
+    };
+
+    unsafe {
+        let data_ptr = Box::into_raw(Box::new(data));
+        let instance_ptr = Box::into_raw(Box::new(gd_instance));
+
+        get_interface().script_instance_create.unwrap()(
+            instance_ptr,
+            data_ptr as GDExtensionScriptInstanceDataPtr,
+        )
+    }
```

Comment:
Missing safety comment here, the unsafe block also seems to cover a bit much, does it really need to cover all these lines?

Also will this leak memory? it isn't clear to me that godot will clean up these pointers anywhere. It could be nice to have a comment stating that godot is now responsible for the memory after this point if that's the case.

Reply:
> the unsafe block also seems to cover a bit much

Right I should reduce that.

> Also will this leak memory?

`free_func` will clean up the memory once Godot calls it. (I will add the comment)